feat(performance-metrics): add RobotContextTracker

DerekMaggio opened the pull request:
# Overview

Create RobotContextTracker class.

This class provides a `track` method. 
The track method should be used as a decorator on either a function/method. It takes a RobotContextState enum value to label what state the RobotContextTracker is tracking. Uses `FunctionTimer` to measure execution time and stores results in a list. 

RobotContextTracker is defaulted not to track anything at all. To turn on tracking, instantiate the class with `should_track=True`. When not tracking, the `track` method calls the underlying wrapped function as quickly as possible.

# Test Plan

- See test_robot_contest_tracker.py

# Changelog

- Add RobotContextTracker class
- Add test_robot_context_tracker.py

# Review requests

None

# Risk assessment

Low, not being used on any production code

DerekMaggio committed:
WIP

DerekMaggio committed:
test: add more test cases

DerekMaggio committed:
chore: don't allow decorator ussage

DerekMaggio committed:
test: allow end times to be the same on nested context managers

DerekMaggio committed:
chore: don't talk so much

DerekMaggio committed:
formatting and linting

DerekMaggio committed:
Starting to address PR feedback

DerekMaggio committed:
Forgot to add begin time

DerekMaggio committed:
cleanup

DerekMaggio committed:
init

DerekMaggio committed:
clean clean clean

DerekMaggio committed:
more test cases

DerekMaggio committed:
oops

DerekMaggio committed:
Don't use function timer as a decorator

DerekMaggio committed:
formatting and linting

DerekMaggio committed:
chore: add ability to turn off tracking

DerekMaggio committed:
rename

DerekMaggio committed:
chore: go back to context manager

DerekMaggio committed:
missed this in the merge

DerekMaggio committed:
chore: remove FunctionTimer

sfoster1 commented on performance-metrics/src/performance_metrics/robot_context_tracker.py:
```diff
@@ -0,0 +1,58 @@
+
+    def __init__(self, should_track: bool = False) -> None:
+        """Initializes the RobotContextTracker with an empty storage list."""
+        self._storage: List[RawContextData] = []
```
use a `deque` for this. python lists are dynamically-allocated memory arrays and any array append may cause a reallocation, which will cause an uneven perf impact. the deque will have more bounded performance, and I don't believe we're ever planning to access items in here by-index anyway.

sfoster1 reviewed (changes requested):
One small change but looking great, love the decorator approach

DerekMaggio committed:
chore: use double ended queue instead of a list

DerekMaggio committed:
chore(performance-metrics): add CI testing and linting (#14877)

# Overview

Add CI testing and lint checks

DerekMaggio committed:
feat(performance-metrics): store to file (#14882)

# Overview

Add functionality to store CSV data in a file

# Test Plan

- Added test cases to verify storing to a file correctly

# Changelog

- Added `store` method to RobotContextTracker
- Added data shaping to RawContextData
- Tests

# Review requests

None

# Risk assessment

Low

sfoster1 reviewed (approved):
Looks good to me!
